File: lambdas/round-ender/handler.go
// Lambda round-ender: invocada por Step Functions cuando expira el timer de una ronda.
//
// CONCEPTO GO — handler sin eventos SQS:
// Esta Lambda NO viene de SQS. Step Functions la invoca directamente como un Task.
// El "evento" de entrada es el estado JSON del state machine (sfnState).
// La función devuelve un nuevo sfnState que Step Functions usa para decidir
// si continuar el loop (más rondas) o terminar (última ronda).
//
// Flujo:
//   1. Lee respuestas de la ronda que acaba de terminar
//   2. Calcula puntos por jugador (basado en velocidad de respuesta)
//   3. Actualiza scores en DynamoDB
//   4. Broadcast ROUND_END (con respuesta correcta y scores actualizados)
//   5a. Si hay más rondas: guarda la siguiente ronda, broadcast ROUND_START,
//       retorna { hasMoreRounds: true, currentRound: N+1 }
//   5b. Si es la última: actualiza sala a "finished", broadcast GAME_END,
//       retorna { hasMoreRounds: false }
package main

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"sync"

	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"

	internaldynamo "github.com/quizarena/internal/dynamo"
	wsevents "github.com/quizarena/internal/events"
	"github.com/quizarena/internal/wsapi"
)

var (
	logger   *slog.Logger
	dbClient *dynamodb.Client
	wsPoster wsapi.Poster
	initOnce sync.Once
	initErr  error
)

func init() {
	logger = slog.New(slog.NewJSONHandler(os.Stdout, nil))
}

// sfnState es el estado que circula dentro del Step Functions state machine.
// Step Functions pasa este struct como input a round-ender, y round-ender
// lo devuelve (posiblemente modificado) para que el Choice state decida qué hacer.
//
// CONCEPTO GO — struct como contrato de API:
// En Python usarías un dict. En Go el struct actúa como contrato tipado:
// si falta un campo o el tipo no coincide, el compilador o el runtime lo detecta.
type sfnState struct {
	RoomID               string `json:"roomId"`
	CurrentRound         int    `json:"currentRound"`
	TotalRounds          int    `json:"totalRounds"`
	RoundDurationSeconds int    `json:"roundDurationSeconds"`
	HasMoreRounds        bool   `json:"hasMoreRounds"`
}

// handler es la función principal. Step Functions espera que devuelva el nuevo estado.
func handler(ctx context.Context, state sfnState) (sfnState, error) {
	if err := ensureClients(ctx); err != nil {
		return sfnState{}, fmt.Errorf("init clients: %w", err)
	}

	logger.InfoContext(ctx, "round ending",
		slog.String("roomId", state.RoomID),
		slog.Int("round", state.CurrentRound),
		slog.Int("total", state.TotalRounds),
	)

	// Paso 1: leer los metadatos de la ronda que acaba de terminar.
	round, err := internaldynamo.GetRound(ctx, dbClient, state.RoomID, state.CurrentRound)
	if err != nil {
		return sfnState{}, fmt.Errorf("get round: %w", err)
	}
	if round == nil {
		return sfnState{}, fmt.Errorf("round %d not found in room %s", state.CurrentRound, state.RoomID)
	}

	// Paso 2: leer todas las respuestas enviadas para esta ronda.
	answers, err := internaldynamo.GetRoundAnswers(ctx, dbClient, state.RoomID, state.CurrentRound)
	if err != nil {
		return sfnState{}, fmt.Errorf("get answers: %w", err)
	}

	// Paso 3: calcular y persistir puntos por jugador.
	for _, ans := range answers {
		isCorrect := ans.Answer == round.CorrectAnswer
		points := internaldynamo.CalcPoints(isCorrect, ans.AnsweredAt, round.StartedAt, round.TimeLimitMs)
		if points > 0 {
			if err := internaldynamo.AddPlayerScore(ctx, dbClient, state.RoomID, ans.PlayerID, points); err != nil {
				// Log el error pero continúa — no queremos abortar la ronda por un jugador.
				logger.ErrorContext(ctx, "failed to add score",
					slog.String("playerId", ans.PlayerID),
					slog.String("error", err.Error()),
				)
			}
		}
	}

	// Paso 4: leer scores actualizados para incluirlos en ROUND_END.
	players, err := internaldynamo.ListPlayersInRoom(ctx, dbClient, state.RoomID)
	if err != nil {
		return sfnState{}, fmt.Errorf("list players: %w", err)
	}

	playerInfos := toPlayerInfos(players)

	// Paso 5: broadcast ROUND_END con respuesta correcta y scores.
	roundEndMsg := wsevents.OutboundMessage{
		Type: wsevents.TypeRoundEnd,
		Payload: wsevents.RoundEndPayload{
			RoundNumber:   state.CurrentRound,
			CorrectAnswer: round.CorrectAnswer,
			Scores:        playerInfos,
		},
	}
	broadcastToAll(ctx, players, roundEndMsg)

	// Paso 6: ¿hay más rondas?
	if state.CurrentRound >= state.TotalRounds {
		// --- Última ronda: terminar el juego ---
		if err := internaldynamo.UpdateRoomStatus(ctx, dbClient, state.RoomID, "finished"); err != nil {
			logger.ErrorContext(ctx, "update room to finished", slog.String("error", err.Error()))
		}

		gameEndMsg := wsevents.OutboundMessage{
			Type: wsevents.TypeGameEnd,
			Payload: wsevents.GameEndPayload{
				Scores: playerInfos,
			},
		}
		broadcastToAll(ctx, players, gameEndMsg)

		logger.InfoContext(ctx, "game finished", slog.String("roomId", state.RoomID))

		// Step Functions: hasMoreRounds=false → el Choice state va a GameComplete (Succeed).
		return sfnState{HasMoreRounds: false}, nil
	}

	// --- Hay más rondas: preparar la siguiente ---
	nextRound := state.CurrentRound + 1

	nextRoundData, err := internaldynamo.GetRound(ctx, dbClient, state.RoomID, nextRound)
	if err != nil {
		return sfnState{}, fmt.Errorf("get next round: %w", err)
	}
	if nextRoundData == nil {
		return sfnState{}, fmt.Errorf("next round %d not found", nextRound)
	}

	// Broadcast ROUND_START con la siguiente pregunta (sin correctAnswer).
	roundStartMsg := wsevents.OutboundMessage{
		Type: wsevents.TypeRoundStart,
		Payload: wsevents.RoundStartPayload{
			RoundNumber: nextRound,
			RoundID:     fmt.Sprintf("ROUND#%03d", nextRound),
			Question: wsevents.QuestionInfo{
				QuestionID: nextRoundData.QuestionID,
				Text:       nextRoundData.Question,
				Options:    nextRoundData.Options,
			},
			TimeLimitMs: nextRoundData.TimeLimitMs,
		},
	}
	broadcastToAll(ctx, players, roundStartMsg)

	logger.InfoContext(ctx, "next round started",
		slog.String("roomId", state.RoomID),
		slog.Int("round", nextRound),
	)

	// Step Functions: hasMoreRounds=true → el Choice state vuelve a WaitForRound.
	// currentRound actualizado para que el próximo ciclo use el número correcto.
	return sfnState{
		RoomID:               state.RoomID,
		CurrentRound:         nextRound,
		TotalRounds:          state.TotalRounds,
		RoundDurationSeconds: nextRoundData.TimeLimitMs / 1000,
		HasMoreRounds:        true,
	}, nil
}

// broadcastToAll envía en paralelo a todos los jugadores.
func broadcastToAll(ctx context.Context, players []internaldynamo.PlayerRecord, msg wsevents.OutboundMessage) {
	var wg sync.WaitGroup
	for _, p := range players {
		wg.Add(1)
		go func(player internaldynamo.PlayerRecord) {
			defer wg.Done()
			if err := wsPoster.PostMessage(ctx, player.ConnectionID, msg); err != nil {
				logger.WarnContext(ctx, "broadcast failed",
					slog.String("connectionId", player.ConnectionID),
					slog.String("error", err.Error()),
				)
			}
		}(p)
	}
	wg.Wait()
}

// toPlayerInfos convierte []PlayerRecord a []PlayerInfo para los mensajes WS.
func toPlayerInfos(records []internaldynamo.PlayerRecord) []wsevents.PlayerInfo {
	infos := make([]wsevents.PlayerInfo, len(records))
	for i, r := range records {
		infos[i] = wsevents.PlayerInfo{
			PlayerID: r.PlayerID,
			Username: r.Username,
			Score:    r.Score,
		}
	}
	return infos
}

func ensureClients(ctx context.Context) error {
	initOnce.Do(func() {
		cfg, err := config.LoadDefaultConfig(ctx)
		if err != nil {
			initErr = err
			return
		}
		dbClient = dynamodb.NewFromConfig(cfg)

		poster, err := wsapi.NewClient(ctx)
		if err != nil {
			initErr = fmt.Errorf("ws client: %w", err)
			return
		}
		wsPoster = poster
	})
	return initErr
}

func main() {
	lambda.Start(handler)
}

File: lambdas/round-ender/handler.py
import json
import logging
import os
from concurrent.futures import ThreadPoolExecutor, as_completed
import boto3

from shared.events.types import (
    TYPE_ROUND_END, TYPE_ROUND_START, TYPE_GAME_END, outbound,
)
from shared.dynamo import room as dynamo_room
from shared.dynamo import game as dynamo_game
from shared.wsapi import post as wsapi

logger = logging.getLogger()
logger.setLevel(logging.INFO)

_db = None
_eb = None
_ws = None


def get_clients():
    global _db, _eb, _ws
    if _db is None:
        _db = boto3.resource("dynamodb")
        _eb = boto3.client("events")
        _ws = wsapi.new_client()
    return _db, _eb, _ws


def broadcast_to_all(ws, players, message):
    if not players:
        return
    with ThreadPoolExecutor(max_workers=max(len(players), 1)) as executor:
        futures = {executor.submit(wsapi.post_message, ws, p["connectionId"], message): p for p in players}
        for future in as_completed(futures):
            try:
                future.result()
            except Exception as e:
                p = futures[future]
                logger.warning(json.dumps({"event": "broadcast-failed", "connectionId": p["connectionId"], "error": str(e)}))


def to_player_infos(players):
    return [{"playerId": p["playerId"], "username": p["username"], "score": int(p.get("score", 0))}
            for p in players]


def publish_game_ended(eb, room_id, players):
    max_score = max((int(p.get("score", 0)) for p in players), default=-1)
    results = [
        {"playerId": p["playerId"], "username": p["username"],
         "score": int(p.get("score", 0)), "isWinner": int(p.get("score", 0)) == max_score and max_score >= 0}
        for p in players
    ]
    detail = json.dumps({"roomId": room_id, "players": results})
    eb.put_events(Entries=[{
        "EventBusName": os.environ["EVENTBRIDGE_BUS_NAME"],
        "Source":       "quizarena",
        "DetailType":   "game.ended",
        "Detail":       detail,
    }])


def handler(event, context):
    """
    Invocada directamente por Step Functions.
    event  = el estado JSON del state machine (sfnState).
    return = nuevo sfnState que el Choice state evalúa.
    """
    db, eb, ws = get_clients()

    room_id       = event["roomId"]
    current_round = event["currentRound"]
    total_rounds  = event["totalRounds"]

    logger.info(json.dumps({"event": "round-ending", "roomId": room_id, "round": current_round, "total": total_rounds}))

    # Idempotency guard: si ya fue procesada (por el trigger temprano), salir sin hacer nada.
    if not dynamo_game.try_mark_round_ended(db, room_id, current_round):
        logger.info(json.dumps({"event": "round-already-ended", "roomId": room_id, "round": current_round}))
        # Devolver el estado que el Choice state necesita para continuar o terminar.
        if current_round >= total_rounds:
            return {"hasMoreRounds": False}
        next_round = current_round + 1
        next_data  = dynamo_game.get_round(db, room_id, next_round)
        if not next_data:
            return {"hasMoreRounds": False}
        return {
            "roomId":               room_id,
            "currentRound":         next_round,
            "totalRounds":          total_rounds,
            "roundDurationSeconds": int(next_data["timeLimitMs"]) // 1000,
            "hasMoreRounds":        True,
        }

    round_data = dynamo_game.get_round(db, room_id, current_round)
    if not round_data:
        raise RuntimeError(f"Ronda {current_round} no encontrada en sala {room_id}")

    answers = dynamo_game.get_round_answers(db, room_id, current_round)

    for ans in answers:
        is_correct = int(ans["answer"]) == int(round_data["correctAnswer"])
        points = dynamo_game.calc_points(
            is_correct,
            int(ans["answeredAt"]),
            int(round_data["startedAt"]),
            int(round_data["timeLimitMs"]),
        )
        if points > 0:
            try:
                dynamo_game.add_player_score(db, room_id, ans["playerId"], points)
            except Exception as e:
                logger.error(json.dumps({"event": "add-score-failed", "playerId": ans["playerId"], "error": str(e)}))

    players      = dynamo_room.list_players_in_room(db, room_id)
    player_infos = to_player_infos(players)

    broadcast_to_all(ws, players, outbound(TYPE_ROUND_END, {
        "roundNumber":   current_round,
        "correctAnswer": int(round_data["correctAnswer"]),
        "scores":        player_infos,
    }))

    if current_round >= total_rounds:
        # Última ronda: terminar el juego
        dynamo_game.update_room_status(db, room_id, "finished")
        broadcast_to_all(ws, players, outbound(TYPE_GAME_END, {"scores": player_infos}))
        logger.info(json.dumps({"event": "game-finished", "roomId": room_id}))

        try:
            publish_game_ended(eb, room_id, players)
        except Exception as e:
            logger.error(json.dumps({"event": "publish-game-ended-failed", "error": str(e)}))

        return {"hasMoreRounds": False}

    next_round = current_round + 1
    next_data  = dynamo_game.get_round(db, room_id, next_round)
    if not next_data:
        raise RuntimeError(f"Siguiente ronda {next_round} no encontrada")

    broadcast_to_all(ws, players, outbound(TYPE_ROUND_START, {
        "roundNumber": next_round,
        "roundId":     f"ROUND#{next_round:03d}",
        "question":    {"questionId": next_data["questionId"], "text": next_data["question"], "options": next_data["options"]},
        "timeLimitMs": int(next_data["timeLimitMs"]),
    }))

    logger.info(json.dumps({"event": "next-round-started", "roomId": room_id, "round": next_round}))

    return {
        "roomId":               room_id,
        "currentRound":         next_round,
        "totalRounds":          total_rounds,
        "roundDurationSeconds": int(next_data["timeLimitMs"]) // 1000,
        "hasMoreRounds":        True,
    }
